File: Problem_1_DP.cpp
//Space Complexity: O(NM), where n is number of houses and m is number of colors we can paint, i.e 3.
//Time Complexity: O(NM), where n is the length of the array.
//Executed successfully  on leetcode.

class Solution {    
public:
    int minCost(vector<vector<int>>& costs) {
        if(costs.size()==0){
            return 0;
        }
        for(int i=1; i< costs.size(); i++){
            costs.at(i).at(0) += min(costs.at(i-1).at(1), costs.at(i-1).at(2));
            costs.at(i).at(1) += min(costs.at(i-1).at(0), costs.at(i-1).at(2));
            costs.at(i).at(2) += min(costs.at(i-1).at(1), costs.at(i-1).at(0));
        }
        return min(costs.at(costs.size()-1).at(0),min(costs.at(costs.size()-1).at(1), costs.at(costs.size()-1).at(2)));
    }
};

File: Problem_1_DP.java
// Time complexity - O(n)
// Space complexity - O(1)

// DP Solution using bottoms up approach.

class Solution {
    public int minCost(int[][] costs) {
        if(costs.length == 0 || costs == null){
            return 0;
        }
        
        for(int i = 1;i< costs.length ;i++){
            costs[i][0] += Math.min(costs[i-1][1],costs[i-1][2]);
            costs[i][1] += Math.min(costs[i-1][0],costs[i-1][2]);
            costs[i][2] += Math.min(costs[i-1][1],costs[i-1][0]);
        }
    
        return Math.min(costs[costs.length-1][0], Math.min(costs[costs.length-1][1],costs[costs.length-1][2]));
    }
    
   
}
